src/java.base/share/classes/java/lang/classfile/AnnotationValue.java

```diff
@@ -79,134 +79,209 @@ sealed interface OfArray extends AnnotationValue
     }
 
     /**
-     * Models a constant-valued element
+     * Models a constant-valued element.
      *
+     * @param <C> the constant pool entry type
+     * @param <R> the resolved live constant type
      * @sealedGraph
      * @since 22
      */
     @PreviewFeature(feature = PreviewFeature.Feature.CLASSFILE_API)
-    sealed interface OfConstant extends AnnotationValue
-            permits AnnotationValue.OfString, AnnotationValue.OfDouble,
-                    AnnotationValue.OfFloat, AnnotationValue.OfLong,
-                    AnnotationValue.OfInteger, AnnotationValue.OfShort,
-                    AnnotationValue.OfCharacter, AnnotationValue.OfByte,
-                    AnnotationValue.OfBoolean, AnnotationImpl.OfConstantImpl {
-        /** {@return the constant} */
-        AnnotationConstantValueEntry constant();
-        /** {@return the constant} */
-        ConstantDesc constantValue();
-    }
-
-    /**
-     * Models a constant-valued element
+    sealed interface OfConstant<C extends AnnotationConstantValueEntry, R extends Comparable<R> & Constable>
```

Comment:
I suggest to minimize use of generics in combination with sealed interface trees and pattern matching. It does not give many benefits, it takes out a part of the code readability and adds many unnecessary brackets and questionmarks: 
```
case AnnotationValue.OfConstant<?, ?> oc ->
```

`<R extends Comparable<R> & Constable>` at least brings a benefit of types combination. However I still think it does not outweigh the generics use cost. None of the generics parameter is near to orthogonal to the `OfConstant` nor critical for the API to operate.

Reply:
Removed the type parameters. Please review again.